### PR TITLE
Dim collapsing reorder pr

### PR DESCRIPTION
**Pull request by jjsjann123**
This PR is a rework of #215 with cherry-picked commit from #225.

The PR utilizes permutation of tensor in integration, as well as passing contiguity information into codegen. This ensures that an input `TensorView` is permuted so that its dimensions go from the slowest to fastest.

Things that are not yet implemented (and will only be handles in this PR):
- only supports point-wise operations. (reduction will require we to map/propagate reduction axes to input)
- requires all inputs to be in the same permuted order. (otherwise, integration cannot permute without changing the semantic meaning of an dimension)
- only support profiling executor. (to support legacy executor, we need to home-brew our own graph caching strategy in order to get a reasonable shape/contiguity information on inputs).

### CODE REVIEW

**Comment by jjsjann123** [comment]
this PR only gets code in functionally. code structure is ugly, it will be cleaned up / refactored in next PR handling graph caching.

@csarofeen for review, you should only look at `manage.cpp` & `tensor_view.cpp`. hopefully that will make review faster.

File: test/cpp/jit/test_gpu.cpp
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp]: Nit: #if 0 may work slightly better (since it minimizes the number of changed lines, so easier to track the history and merges)

File: test/cpp/jit/test_gpu.cpp
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:2937]: ditto

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/arith.cpp:87]: Not a big deal, but std::vector<bool> has a number of issues that makes it worth avoiding when possible. 

Would it be too much work to use std::bitset as an alternative?

File: torch/csrc/jit/codegen/cuda/manager.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/manager.cpp:296]: Does this support reduction? Would assume there'd be similar issues with sorting.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: naming convention: use `cmi` or `contig_merge_info`

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: naming: `meargable_info` 

also, `const` ?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: rename `contig3` -> `result` ?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:58]: This should be fixed now in https://github.com/csarofeen/pytorch/pull/225 but will include this and other fixes there after we merge this.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:530]: see earlier comment re. `vector<bool>` vs `bitset`.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:33]: nit: empty line here

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:42]: const?

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: const ?

File: torch/csrc/jit/codegen/cuda/manager.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/manager.cpp:296]: Same restriction applies and we are excluding graph with reduction operation for now.

File: torch/csrc/jit/codegen/cuda/arith.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/arith.cpp:87]: it would take quite a bit of work where we'd have to specialize it on all number of dimensions we expect in root domain.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:33]: No it's not an empty line, but format by github.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:42]: This one is not applicable as `overall_contiguity` is changed/updated later.

**Comment by jjsjann123** [comment]
I believe I address most review comments.

The comment on bitset vs std::vector is being postponed in offline discussion due to the requirement of compile time size.
I'll merge this PR once clang-tidy passes.

**Pull request merged**
